File: backend/app/services/executive_intelligence_service.py
"""
Executive Intelligence Service

Calculates executive-level metrics and investment recommendations from market intelligence data.
Provides actionable insights for C-level decision making.
"""

import logging
from typing import Dict, Any, Optional, List
from datetime import datetime

logger = logging.getLogger(__name__)


class ExecutiveIntelligenceService:
    """Service for calculating executive-level market intelligence metrics."""

    def calculate_location_score(self, location_intel: Optional[Dict], demographics: Optional[Dict]) -> Optional[float]:
        """
        Calculate location quality score (0-10).

        Factors:
        - Walk/Bike/Transit scores (30%)
        - School ratings (20%)
        - Crime index (20%)
        - Amenities proximity (15%)
        - Demographics (15%)
        """
        if not location_intel or not location_intel.get('data'):
            return None

        try:
            data = location_intel['data']
            score = 0.0

            # Mobility scores (30%) - normalized to 0-3
            walk_score = data.get('walk_score', 0) / 100 * 1.2
            bike_score = data.get('bike_score', 0) / 100 * 0.9
            transit_score = data.get('transit_score', 0) / 100 * 0.9
            score += (walk_score + bike_score + transit_score)

            # School ratings (20%) - normalized to 0-2
            school_rating = data.get('school_rating_avg', 5) / 10 * 2.0
            score += school_rating

            # Crime index (20%) - inverted (lower crime = higher score) to 0-2
            crime_index = data.get('crime_index', 50)
            crime_score = (100 - crime_index) / 100 * 2.0
            score += crime_score

            # Amenities (15%) - normalized to 0-1.5
            amenities = data.get('amenities', {})
            amenity_score = 0
            if amenities.get('grocery_stores_1mi', 0) > 0:
                amenity_score += 0.4
            if amenities.get('restaurants_1mi', 0) > 3:
                amenity_score += 0.3
            if amenities.get('parks_1mi', 0) > 0:
                amenity_score += 0.3
            if amenities.get('schools_2mi', 0) > 0:
                amenity_score += 0.3
            if amenities.get('hospitals_5mi', 0) > 0:
                amenity_score += 0.2
            score += min(amenity_score, 1.5)

            # Demographics quality (15%) - normalized to 0-1.5
            if demographics and demographics.get('data'):
                demo = demographics['data']
                income_score = 0
                edu_score = 0

                median_income = demo.get('median_household_income', 0)
                if median_income > 80000:
                    income_score = 0.8
                elif median_income > 60000:
                    income_score = 0.6
                elif median_income > 40000:
                    income_score = 0.4

                college_pct = demo.get('college_educated_pct', 0)
                if college_pct > 40:
                    edu_score = 0.7
                elif college_pct > 25:
                    edu_score = 0.5
                elif college_pct > 15:
                    edu_score = 0.3

                score += (income_score + edu_score)

            # Cap at 10
            return round(min(score, 10.0), 1)

        except Exception as e:
            logger.error(f"Error calculating location score: {e}")
            return None

    def calculate_market_cap_rate(self, competitive_analysis: Optional[Dict]) -> float:
        """Calculate market average cap rate from competitive analysis or economic data."""
        if not competitive_analysis or not competitive_analysis.get('data'):
            return 0.0

        try:
            data = competitive_analysis['data']
            avg_cap = data.get('market_avg_cap_rate', 0)
            return round(avg_cap, 2) if avg_cap else 0.0
        except Exception as e:
            logger.error(f"Error calculating market cap rate: {e}")
            return 0.0

    def calculate_rent_growth(self, forecasts: Optional[Dict]) -> Optional[float]:
        """Extract rent growth forecast from forecasts data."""
        if not forecasts or not forecasts.get('data'):
            return None

        try:
            data = forecasts['data']
            rent_growth = data.get('rent_growth_3yr_cagr', None)
            return round(rent_growth, 1) if rent_growth else None
        except Exception as e:
            logger.error(f"Error extracting rent growth: {e}")
            return None

    def calculate_risk_score(self,
                            esg_assessment: Optional[Dict],
                            economic_indicators: Optional[Dict],
                            forecasts: Optional[Dict],
                            location_intel: Optional[Dict]) -> int:
        """
        Calculate composite risk score (0-100, higher = more risk).

        Factors:
        - ESG risks (30%)
        - Economic conditions (25%)
        - Market forecast risks (25%)
        - Location/crime (20%)
        """
        try:
            risk = 0

            # ESG Risks (30 points max)
            if esg_assessment and esg_assessment.get('data'):
                esg_data = esg_assessment['data']
                esg_score = esg_data.get('composite_score', {}).get('overall', {}).get('score', 75)
                # Lower ESG score = higher risk (invert)
                esg_risk = (100 - esg_score) / 100 * 30
                risk += esg_risk
            else:
                risk += 15  # Moderate risk if no ESG data

            # Economic Conditions (25 points max)
            if economic_indicators and economic_indicators.get('data'):
                econ_data = economic_indicators['data']

                # Recession probability
                recession_prob = econ_data.get('recession_probability', {}).get('value', 0)
                risk += (recession_prob / 100) * 10

                # Unemployment
                unemployment = econ_data.get('unemployment_rate', {}).get('value', 4)
                if unemployment > 6:
                    risk += 8
                elif unemployment > 5:
                    risk += 5
                elif unemployment > 4:
                    risk += 2

                # Mortgage rates
                mortgage_rate = econ_data.get('mortgage_rate_30y', {}).get('value', 6)
                if mortgage_rate > 7:
                    risk += 7
                elif mortgage_rate > 6:
                    risk += 4
            else:
                risk += 12  # Moderate risk if no economic data

            # Market Forecast Risks (25 points max)
            if forecasts and forecasts.get('data'):
                forecast_data = forecasts['data']

                # Rent growth (negative growth = risk)
                rent_growth = forecast_data.get('rent_growth_3yr_cagr', 3)
                if rent_growth < 0:
                    risk += 15
                elif rent_growth < 2:
                    risk += 10
                elif rent_growth < 3:
                    risk += 5

                # Vacancy risk
                vacancy_forecast = forecast_data.get('vacancy_rate_forecast', 5)
                if vacancy_forecast > 10:
                    risk += 10
                elif vacancy_forecast > 7:
                    risk += 5
            else:
                risk += 12  # Moderate risk if no forecast

            # Location/Crime (20 points max)
            if location_intel and location_intel.get('data'):
                loc_data = location_intel['data']
                crime_index = loc_data.get('crime_index', 50)
                # Higher crime = higher risk
                risk += (crime_index / 100) * 20
            else:
                risk += 10  # Moderate risk if no location data

            return round(min(risk, 100))

        except Exception as e:
            logger.error(f"Error calculating risk score: {e}")
            return 50  # Default moderate risk

    def calculate_opportunity_score(self,
                                   location_score: Optional[float],
                                   rent_growth: Optional[float],
                                   demographics: Optional[Dict],
                                   competitive_analysis: Optional[Dict]) -> int:
        """
        Calculate opportunity score (0-100, higher = better opportunity).

        Factors:
        - Rent growth potential (30%)
        - Location quality (25%)
        - Demographic trends (25%)
        - Competitive positioning (20%)
        """
        try:
            score = 0

            # Rent Growth Potential (30 points)
            if rent_growth is not None:
                if rent_growth >= 5:
                    score += 30
                elif rent_growth >= 3:
                    score += 20
                elif rent_growth >= 2:
                    score += 15
                elif rent_growth >= 0:
                    score += 10
            else:
                score += 15  # Moderate if unknown

            # Location Quality (25 points)
            if location_score is not None:
                score += (location_score / 10) * 25
            else:
                score += 12  # Moderate if unknown

            # Demographics (25 points)
            if demographics and demographics.get('data'):
                demo = demographics['data']

                # Income growth indicator
                median_income = demo.get('median_household_income', 0)
                if median_income > 80000:
                    score += 12
                elif median_income > 60000:
                    score += 8
                elif median_income > 40000:
                    score += 5

                # Education (proxy for growth)
                college_pct = demo.get('college_educated_pct', 0)
                if college_pct > 40:
                    score += 10
                elif college_pct > 25:
                    score += 6
                elif college_pct > 15:
                    score += 3

                # Population density (demand indicator)
                population = demo.get('population', 0)
                if population > 5000:
                    score += 3
            else:
                score += 12  # Moderate if unknown

            # Competitive Positioning (20 points)
            if competitive_analysis and competitive_analysis.get('data'):
                comp_data = competitive_analysis['data']

                # New supply vs demand
                absorption = comp_data.get('absorption_rate_units_per_mo', 0)
                new_supply = comp_data.get('new_supply_pipeline_units', 0)

                if absorption > 0 and new_supply / max(absorption, 1) < 12:
                    # Less than 1 year of supply = opportunity
                    score += 15
                elif absorption > 0 and new_supply / max(absorption, 1) < 24:
                    score += 10
                else:
                    score += 5

                # Market cap rate vs trend
                score += 5  # Placeholder
            else:
                score += 10  # Moderate if unknown

            return round(min(score, 100))

        except Exception as e:
            logger.error(f"Error calculating opportunity score: {e}")
            return 50  # Default moderate

    def generate_investment_recommendation(self,
                                         risk_score: int,
                                         opportunity_score: int,
                                         your_cap_rate: float,
                                         market_cap_rate: float) -> Dict[str, Any]:
        """
        Generate executive investment recommendation.

        Returns:
        - action: "BUY", "HOLD", "SELL", "REVIEW"
        - confidence: 0-100
        - rationale: List of key reasons
        - priority: "HIGH", "MEDIUM", "LOW"
        """
        try:
            # Calculate decision matrix score
            # Opportunity - Risk = Net Attractiveness
            net_score = opportunity_score - risk_score

            # Cap rate comparison
            cap_rate_premium = your_cap_rate - market_cap_rate if market_cap_rate > 0 else 0

            # Decision logic
            action = "REVIEW"
            confidence = 50
            rationale = []
            priority = "MEDIUM"

            # Strong Buy Signals
            if net_score >= 30 and opportunity_score >= 70:
                action = "BUY"
                confidence = 85
                priority = "HIGH"
                rationale.append(f"Exceptional opportunity (score: {opportunity_score}/100)")
                rationale.append(f"Risk is manageable ({risk_score}/100)")

            # Moderate Buy Signals
            elif net_score >= 15 and opportunity_score >= 60:
                action = "BUY"
                confidence = 70
                priority = "MEDIUM"
                rationale.append(f"Strong opportunity (score: {opportunity_score}/100)")
                rationale.append(f"Acceptable risk level ({risk_score}/100)")

            # Hold Signals - performing well
            elif net_score >= 0 and cap_rate_premium >= 0:
                action = "HOLD"
                confidence = 75
                priority = "LOW"
                rationale.append(f"Asset performing at/above market (cap rate +{cap_rate_premium:.1f}%)")
                rationale.append(f"Balanced opportunity/risk profile")

            # Hold Signals - neutral
            elif net_score >= -15:
                action = "HOLD"
                confidence = 60
                priority = "MEDIUM"
                rationale.append(f"Market conditions stable")
                rationale.append(f"Monitor for changes in risk/opportunity balance")

            # Sell Signals - high risk
            elif risk_score >= 70:
                action = "SELL"
                confidence = 80
                priority = "HIGH"
                rationale.append(f"High risk exposure ({risk_score}/100)")
                rationale.append(f"Limited upside potential ({opportunity_score}/100)")

            # Sell Signals - underperforming
            elif cap_rate_premium < -1.0 and opportunity_score < 40:
                action = "SELL"
                confidence = 75
                priority = "HIGH"
                rationale.append(f"Underperforming market (cap rate {cap_rate_premium:.1f}%)")
                rationale.append(f"Weak growth prospects")

            # Review Signals
            else:
                action = "REVIEW"
                confidence = 50
                priority = "MEDIUM"
                rationale.append(f"Mixed signals - detailed analysis recommended")
                rationale.append(f"Opportunity: {opportunity_score}/100, Risk: {risk_score}/100")

            # Add cap rate context if available
            if market_cap_rate > 0:
                if cap_rate_premium >= 0.5:
                    rationale.append(f"Outperforming market by {cap_rate_premium:.1f}%")
                elif cap_rate_premium <= -0.5:
                    rationale.append(f"Underperforming market by {abs(cap_rate_premium):.1f}%")

            return {
                "action": action,
                "confidence": confidence,
                "priority": priority,
                "rationale": rationale,
                "metrics": {
                    "net_score": net_score,
                    "opportunity_score": opportunity_score,
                    "risk_score": risk_score,
                    "cap_rate_premium": round(cap_rate_premium, 2)
                }
            }

        except Exception as e:
            logger.error(f"Error generating investment recommendation: {e}")
            return {
                "action": "REVIEW",
                "confidence": 0,
                "priority": "HIGH",
                "rationale": ["Error calculating recommendation - manual review required"],
                "metrics": {}
            }

    def generate_executive_summary(self, market_intelligence: Dict[str, Any], your_cap_rate: float = 0.0) -> Dict[str, Any]:
        """
        Generate complete executive summary with all calculated metrics.

        Args:
            market_intelligence: Complete market intelligence data
            your_cap_rate: Property's actual cap rate

        Returns:
            Executive summary with scores, recommendation, and key insights
        """
        try:
            # Extract data
            demographics = market_intelligence.get('demographics')
            economic_indicators = market_intelligence.get('economic_indicators')
            location_intelligence = market_intelligence.get('location_intelligence')
            esg_assessment = market_intelligence.get('esg_assessment')
            forecasts = market_intelligence.get('forecasts')
            competitive_analysis = market_intelligence.get('competitive_analysis')
            ai_insights = market_intelligence.get('ai_insights')

            # Calculate scores
            location_score = self.calculate_location_score(location_intelligence, demographics)
            market_cap_rate = self.calculate_market_cap_rate(competitive_analysis)
            rent_growth = self.calculate_rent_growth(forecasts)
            risk_score = self.calculate_risk_score(esg_assessment, economic_indicators, forecasts, location_intelligence)
            opportunity_score = self.calculate_opportunity_score(location_score, rent_growth, demographics, competitive_analysis)

            # Generate recommendation
            recommendation = self.generate_investment_recommendation(
                risk_score, opportunity_score, your_cap_rate, market_cap_rate
            )

            # Extract key insights
            key_findings = []
            if ai_insights and ai_insights.get('data'):
                ai_data = ai_insights['data']
                opportunities = ai_data.get('opportunities', [])
                key_findings.extend(opportunities[:2])  # Top 2

            return {
                "location_score": location_score,
                "market_cap_rate": market_cap_rate,
                "rent_growth": rent_growth,
                "risk_score": risk_score,
                "opportunity_score": opportunity_score,
                "investment_recommendation": recommendation,
                "key_findings": key_findings,
                "executive_summary": {
                    "headline": self._generate_headline(recommendation, location_score, rent_growth),
                    "quick_stats": {
                        "location_quality": self._score_to_label(location_score, "location"),
                        "growth_potential": self._score_to_label(rent_growth, "growth"),
                        "risk_level": self._score_to_label(risk_score, "risk"),
                        "opportunity_level": self._score_to_label(opportunity_score, "opportunity")
                    }
                }
            }

        except Exception as e:
            logger.error(f"Error generating executive summary: {e}")
            return {
                "location_score": None,
                "market_cap_rate": 0.0,
                "rent_growth": None,
                "risk_score": 50,
                "opportunity_score": 50,
                "investment_recommendation": {
                    "action": "REVIEW",
                    "confidence": 0,
                    "priority": "HIGH",
                    "rationale": ["Error generating summary - manual review required"],
                    "metrics": {}
                },
                "key_findings": [],
                "executive_summary": {
                    "headline": "Insufficient data for recommendation",
                    "quick_stats": {}
                }
            }

    def _generate_headline(self, recommendation: Dict, location_score: Optional[float], rent_growth: Optional[float]) -> str:
        """Generate executive headline summary."""
        action = recommendation.get('action', 'REVIEW')
        confidence = recommendation.get('confidence', 0)

        if action == "BUY" and confidence >= 80:
            return f"STRONG BUY: Premium location ({location_score}/10) with {rent_growth}% growth potential"
        elif action == "BUY":
            return f"BUY: Attractive opportunity with manageable risk"
        elif action == "HOLD":
            return f"HOLD: Asset performing at market - monitor conditions"
        elif action == "SELL":
            return f"SELL RECOMMENDED: High risk or underperformance detected"
        else:
            return f"REVIEW REQUIRED: Mixed signals - detailed analysis needed"

    def _score_to_label(self, score: Optional[float], metric_type: str) -> str:
        """Convert numeric score to executive label."""
        if score is None:
            return "Unknown"

        if metric_type == "location":
            if score >= 8:
                return "Premium"
            elif score >= 6:
                return "Strong"
            elif score >= 4:
                return "Average"
            else:
                return "Below Average"

        elif metric_type == "growth":
            if score >= 5:
                return "Exceptional"
            elif score >= 3:
                return "Strong"
            elif score >= 2:
                return "Moderate"
            elif score >= 0:
                return "Stable"
            else:
                return "Declining"

        elif metric_type == "risk":
            if score >= 70:
                return "High Risk"
            elif score >= 50:
                return "Moderate Risk"
            elif score >= 30:
                return "Low Risk"
            else:
                return "Minimal Risk"

        elif metric_type == "opportunity":
            if score >= 75:
                return "Exceptional"
            elif score >= 60:
                return "Strong"
            elif score >= 40:
                return "Moderate"
            else:
                return "Limited"

        return "Unknown"

File: src/services/executiveIntelligenceService.ts
/**
 * Executive Intelligence Service (Frontend)
 *
 * Calculates executive-level metrics and investment recommendations from market intelligence data.
 * Provides actionable insights for C-level decision making.
 */

export interface ExecutiveSummary {
  locationScore: number | null;
  marketCapRate: number;
  rentGrowth: number | null;
  riskScore: number;
  opportunityScore: number;
  investmentRecommendation: InvestmentRecommendation;
  keyFindings: string[];
  executiveSummary: {
    headline: string;
    quickStats: {
      locationQuality: string;
      growthPotential: string;
      riskLevel: string;
      opportunityLevel: string;
    };
  };
}

export interface InvestmentRecommendation {
  action: 'BUY' | 'HOLD' | 'SELL' | 'REVIEW';
  confidence: number;
  priority: 'HIGH' | 'MEDIUM' | 'LOW';
  rationale: string[];
  metrics: {
    netScore: number;
    capRatePremium: number;
  };
}

/**
 * Calculate location quality score (0-10)
 */
export function calculateLocationScore(
  locationIntel: any,
  demographics: any
): number | null {
  if (!locationIntel?.data) return null;

  try {
    const loc = locationIntel.data;
    let score = 0;

    // Mobility scores (30% weight)
    score += (loc.walk_score || 0) / 100 * 1.2;
    score += (loc.bike_score || 0) / 100 * 0.9;
    score += (loc.transit_score || 0) / 100 * 0.9;

    // Schools (20% weight)
    score += (loc.school_rating_avg || 5) / 10 * 2.0;

    // Safety (20% weight) - inverted (lower crime = higher score)
    score += (100 - (loc.crime_index || 50)) / 100 * 2.0;

    // Amenities (15% weight)
    const amenities = loc.amenities || {};
    let amenityScore = 0;
    if ((amenities.grocery_stores_1mi || 0) > 0) amenityScore += 0.4;
    if ((amenities.restaurants_1mi || 0) > 3) amenityScore += 0.3;
    if ((amenities.parks_1mi || 0) > 0) amenityScore += 0.3;
    if ((amenities.schools_2mi || 0) > 0) amenityScore += 0.3;
    if ((amenities.hospitals_5mi || 0) > 0) amenityScore += 0.2;
    score += Math.min(amenityScore, 1.5);

    // Demographics boost (15% weight)
    if (demographics?.data) {
      const demo = demographics.data;
      const income = demo.median_household_income || 0;
      const education = demo.college_educated_pct || 0;

      if (income > 80000) score += 0.8;
      else if (income > 60000) score += 0.6;
      else if (income > 40000) score += 0.4;

      if (education > 40) score += 0.7;
      else if (education > 25) score += 0.5;
      else if (education > 15) score += 0.3;
    }

    return Math.round(Math.min(score, 10) * 10) / 10;
  } catch (error) {
    console.error('Error calculating location score:', error);
    return null;
  }
}

/**
 * Calculate market cap rate from competitive analysis
 */
export function calculateMarketCapRate(competitiveAnalysis: any): number {
  if (!competitiveAnalysis?.data) return 0;
  return competitiveAnalysis.data.market_avg_cap_rate || 0;
}

/**
 * Extract rent growth forecast
 */
export function calculateRentGrowth(forecasts: any): number | null {
  if (!forecasts?.data) return null;
  const growth = forecasts.data.rent_growth_3yr_cagr;
  return growth !== null && growth !== undefined ? Math.round(growth * 10) / 10 : null;
}

/**
 * Calculate risk score (0-100, higher = more risk)
 */
export function calculateRiskScore(
  esgAssessment: any,
  economicIndicators: any,
  forecasts: any,
  locationIntel: any
): number {
  let risk = 0;

  try {
    // ESG Risks (30 points max)
    if (esgAssessment?.data) {
      const esgScore = esgAssessment.data.composite_score?.overall?.score || 75;
      risk += (100 - esgScore) / 100 * 30;
    } else {
      risk += 15; // Moderate risk if no ESG data
    }

    // Economic Conditions (25 points max)
    if (economicIndicators?.data) {
      const econ = economicIndicators.data;

      // Recession probability
      const recessionProb = econ.recession_probability?.value || 0;
      risk += (recessionProb / 100) * 10;

      // Unemployment
      const unemployment = econ.unemployment_rate?.value || 4;
      if (unemployment > 6) risk += 8;
      else if (unemployment > 5) risk += 5;
      else if (unemployment > 4) risk += 2;

      // Mortgage rates
      const mortgageRate = econ.mortgage_rate_30y?.value || 6;
      if (mortgageRate > 7) risk += 7;
      else if (mortgageRate > 6) risk += 4;
    } else {
      risk += 12; // Moderate risk if no economic data
    }

    // Market Forecast Risks (25 points max)
    if (forecasts?.data) {
      const rentGrowth = forecasts.data.rent_growth_3yr_cagr || 0;
      if (rentGrowth < 0) risk += 15;
      else if (rentGrowth < 2) risk += 10;
      else if (rentGrowth < 3) risk += 5;

      const vacancyForecast = forecasts.data.vacancy_rate_forecast || 5;
      if (vacancyForecast > 10) risk += 10;
      else if (vacancyForecast > 7) risk += 5;
    } else {
      risk += 12; // Moderate risk if no forecast
    }

    // Location/Crime (20 points max)
    if (locationIntel?.data) {
      const crimeIndex = locationIntel.data.crime_index || 50;
      risk += (crimeIndex / 100) * 20;
    } else {
      risk += 10; // Moderate risk if no location data
    }

    return Math.round(Math.min(risk, 100));
  } catch (error) {
    console.error('Error calculating risk score:', error);
    return 50; // Default moderate risk
  }
}

/**
 * Calculate opportunity score (0-100, higher = better)
 */
export function calculateOpportunityScore(
  locationScore: number | null,
  rentGrowth: number | null,
  demographics: any,
  competitiveAnalysis: any
): number {
  let score = 0;

  try {
    // Rent Growth Potential (30 points)
    if (rentGrowth !== null) {
      if (rentGrowth >= 5) score += 30;
      else if (rentGrowth >= 3) score += 20;
      else if (rentGrowth >= 2) score += 15;
      else if (rentGrowth >= 0) score += 10;
    } else {
      score += 15; // Moderate if unknown
    }

    // Location Quality (25 points)
    if (locationScore !== null) {
      score += (locationScore / 10) * 25;
    } else {
      score += 12; // Moderate if unknown
    }

    // Demographics (25 points)
    if (demographics?.data) {
      const demo = demographics.data;
      const income = demo.median_household_income || 0;
      const education = demo.college_educated_pct || 0;
      const population = demo.population || 0;

      if (income > 80000) score += 12;
      else if (income > 60000) score += 8;
      else if (income > 40000) score += 5;

      if (education > 40) score += 10;
      else if (education > 25) score += 6;
      else if (education > 15) score += 3;

      if (population > 5000) score += 3;
    } else {
      score += 12; // Moderate if unknown
    }

    // Competitive Positioning (20 points)
    if (competitiveAnalysis?.data) {
      const comp = competitiveAnalysis.data;
      const absorption = comp.absorption_rate_units_per_mo || 0;
      const newSupply = comp.new_supply_pipeline_units || 0;

      if (absorption > 0) {
        const monthsOfSupply = newSupply / Math.max(absorption, 1);
        if (monthsOfSupply < 12) score += 15; // Less than 1 year = strong opportunity
        else if (monthsOfSupply < 24) score += 10;
        else score += 5;
      }

      score += 5; // Additional competitive positioning points
    } else {
      score += 10; // Moderate if unknown
    }

    return Math.round(Math.min(score, 100));
  } catch (error) {
    console.error('Error calculating opportunity score:', error);
    return 50; // Default moderate
  }
}

/**
 * Generate investment recommendation
 */
export function generateInvestmentRecommendation(
  riskScore: number,
  opportunityScore: number,
  yourCapRate: number,
  marketCapRate: number
): InvestmentRecommendation {
  const netScore = opportunityScore - riskScore;
  const capPremium = marketCapRate > 0 ? yourCapRate - marketCapRate : 0;

  let action: 'BUY' | 'HOLD' | 'SELL' | 'REVIEW' = 'REVIEW';
  let confidence = 50;
  let priority: 'HIGH' | 'MEDIUM' | 'LOW' = 'MEDIUM';
  const rationale: string[] = [];

  // Strong Buy
  if (netScore >= 30 && opportunityScore >= 70) {
    action = 'BUY';
    confidence = 85;
    priority = 'HIGH';
    rationale.push(`Exceptional opportunity (${opportunityScore}/100)`);
    rationale.push(`Risk is manageable (${riskScore}/100)`);
  }
  // Moderate Buy
  else if (netScore >= 15 && opportunityScore >= 60) {
    action = 'BUY';
    confidence = 70;
    priority = 'MEDIUM';
    rationale.push(`Strong opportunity (${opportunityScore}/100)`);
    rationale.push(`Acceptable risk level (${riskScore}/100)`);
  }
  // Hold - performing well
  else if (netScore >= 0 && capPremium >= 0) {
    action = 'HOLD';
    confidence = 75;
    priority = 'LOW';
    rationale.push(`Asset performing at/above market (${capPremium >= 0 ? '+' : ''}${capPremium.toFixed(1)}% cap rate)`);
    rationale.push('Balanced opportunity/risk profile');
  }
  // Hold - neutral
  else if (netScore >= -15) {
    action = 'HOLD';
    confidence = 60;
    priority = 'MEDIUM';
    rationale.push('Market conditions stable');
    rationale.push('Monitor for changes in risk/opportunity balance');
  }
  // Sell - high risk
  else if (riskScore >= 70) {
    action = 'SELL';
    confidence = 80;
    priority = 'HIGH';
    rationale.push(`High risk exposure (${riskScore}/100)`);
    rationale.push(`Limited upside potential (${opportunityScore}/100)`);
  }
  // Sell - underperforming
  else if (capPremium < -1.0 && opportunityScore < 40) {
    action = 'SELL';
    confidence = 75;
    priority = 'HIGH';
    rationale.push(`Underperforming market (${capPremium.toFixed(1)}% cap rate)`);
    rationale.push('Weak growth prospects');
  }
  // Review
  else {
    action = 'REVIEW';
    confidence = 50;
    priority = 'MEDIUM';
    rationale.push('Mixed signals - detailed analysis recommended');
    rationale.push(`Opportunity: ${opportunityScore}/100, Risk: ${riskScore}/100`);
  }

  // Add cap rate context if available
  if (marketCapRate > 0 && !rationale.some(r => r.includes('cap rate'))) {
    if (capPremium >= 0.5) {
      rationale.push(`Outperforming market by ${capPremium.toFixed(1)}%`);
    } else if (capPremium <= -0.5) {
      rationale.push(`Underperforming market by ${Math.abs(capPremium).toFixed(1)}%`);
    }
  }

  return {
    action,
    confidence,
    priority,
    rationale,
    metrics: {
      netScore,
      capRatePremium: Math.round(capPremium * 100) / 100
    }
  };
}

/**
 * Generate complete executive summary
 */
export function generateExecutiveSummary(
  marketIntelligence: any,
  yourCapRate: number = 0
): ExecutiveSummary {
  try {
    const { demographics, location_intelligence, esg_assessment, forecasts, competitive_analysis, ai_insights } = marketIntelligence;

    // Calculate all metrics
    const locationScore = calculateLocationScore(location_intelligence, demographics);
    const marketCapRate = calculateMarketCapRate(competitive_analysis);
    const rentGrowth = calculateRentGrowth(forecasts);
    const riskScore = calculateRiskScore(esg_assessment, marketIntelligence.economic_indicators, forecasts, location_intelligence);
    const opportunityScore = calculateOpportunityScore(locationScore, rentGrowth, demographics, competitive_analysis);
    const investmentRecommendation = generateInvestmentRecommendation(riskScore, opportunityScore, yourCapRate, marketCapRate);
    const safeLocationScore = locationScore ?? 0;
    const safeRentGrowth = rentGrowth ?? 0;

    // Extract key findings
    const keyFindings: string[] = [];
    if (ai_insights?.data) {
      const opportunities = ai_insights.data.opportunities || [];
      keyFindings.push(...opportunities.slice(0, 2));
    }

    // Generate headline
    const { action, rationale } = investmentRecommendation;
    let headline = `${action}: ${rationale[0] || 'Evaluation needed'}`;
    if (action === 'BUY' && investmentRecommendation.confidence >= 80 && locationScore !== null) {
      headline = `STRONG BUY: Premium location (${safeLocationScore}/10) with ${safeRentGrowth}% growth potential`;
    }

    return {
      locationScore,
      marketCapRate,
      rentGrowth,
      riskScore,
      opportunityScore,
      investmentRecommendation,
      keyFindings,
      executiveSummary: {
        headline,
        quickStats: {
          locationQuality: safeLocationScore >= 8 ? 'Premium' : safeLocationScore >= 6 ? 'Strong' : safeLocationScore >= 4 ? 'Average' : 'Below Average',
          growthPotential: safeRentGrowth >= 5 ? 'Exceptional' : safeRentGrowth >= 3 ? 'Strong' : safeRentGrowth >= 2 ? 'Moderate' : safeRentGrowth >= 0 ? 'Stable' : 'Declining',
          riskLevel: riskScore >= 70 ? 'High Risk' : riskScore >= 50 ? 'Moderate Risk' : riskScore >= 30 ? 'Low Risk' : 'Minimal Risk',
          opportunityLevel: opportunityScore >= 75 ? 'Exceptional' : opportunityScore >= 60 ? 'Strong' : opportunityScore >= 40 ? 'Moderate' : 'Limited'
        }
      }
    };
  } catch (error) {
    console.error('Error generating executive summary:', error);
    return {
      locationScore: null,
      marketCapRate: 0,
      rentGrowth: null,
      riskScore: 50,
      opportunityScore: 50,
      investmentRecommendation: {
        action: 'REVIEW',
        confidence: 0,
        priority: 'HIGH',
        rationale: ['Error calculating recommendation - manual review required'],
        metrics: { netScore: 0, capRatePremium: 0 }
      },
      keyFindings: [],
      executiveSummary: {
        headline: 'Insufficient data for recommendation',
        quickStats: {
          locationQuality: 'Unknown',
          growthPotential: 'Unknown',
          riskLevel: 'Unknown',
          opportunityLevel: 'Unknown'
        }
      }
    };
  }
}
